File: java/jpa-backend/src/main/java/chaincue/tech/jpabackend/masters/course_master/CourseHelper.java
package chaincue.tech.jpabackend.masters.course_master;

import chaincue.tech.jpabackend.masters.student_master.StudentRepository;
import chaincue.tech.jpabackend.masters.teacher_master.Teacher;
import chaincue.tech.jpabackend.masters.teacher_master.TeacherRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class CourseHelper {
    private final CourseRepository courseRepository;
    private final TeacherRepository teacherRepository;
    private final StudentRepository studentRepository;

    public Course saveCourse(String name) {
        Course course = Course.createCourse(name);
        return courseRepository.save(course);
    }

    @Transactional
    public Course addTeacherToCourse(String courseId, String teacherId) {
        Optional<Course> optionalCourse = courseRepository.findById(courseId);
        Optional<Teacher> optionalTeacher = teacherRepository.findById(teacherId);

        if(!optionalCourse.isPresent() || !optionalTeacher.isPresent()) {
            throw new RuntimeException("Course or Teacher not found.");
        }

        Course course = optionalCourse.get();
        Teacher teacher = optionalTeacher.get();

        course.getTeacherCourseRelations().add(teacher);
        teacher.getTeacherCourseRelations().add(course);

        courseRepository.save(course);
        teacherRepository.save(teacher);

        return course;
    }

    @Transactional
    public Course addStudentToCourse(String courseId, String teacherId) {
        var course = courseRepository.findById(courseId).orElseThrow(() -> new RuntimeException("Course or Teacher not found."));
        var student = studentRepository.findById(teacherId).orElseThrow(() -> new RuntimeException("Course or Teacher not found."));

        course.getStudentCourseRelations().add(student);
        student.getStudentCourseRelations().add(course);

        courseRepository.save(course);
        studentRepository.save(student);

        return course;
    }

    public List<Course> allCourses() {
        return courseRepository.findAll();
    }

    public Optional<Course> getCourseById(String courseId) {
        return courseRepository.findById(courseId);
    }
}

File: kotlin/jpa-backend/src/main/kotlin/chaincue/tech/jpabackend/masters/course_master/CourseHelper.kt
package chaincue.tech.jpabackend.masters.student_master

import chaincue.tech.jpabackend.masters.teacher_master.Course
import chaincue.tech.jpabackend.masters.teacher_master.TeacherRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class CourseHelper(
        private val courseRepository: CourseRepository,
        private val teacherRepository: TeacherRepository,
        private val studentRepository: StudentRepository
) {
    fun save(username: String): Course {
        val course = Course.create(username)
        return courseRepository.save(course)
    }

    fun find(id: String): Optional<Course> = courseRepository.findById(id)

    @Transactional
    fun addTeacherToCourse(courseId: String, teacherId: String): Course {
        val course = courseRepository.findById(courseId).orElseThrow { RuntimeException("Course or Teacher not found.") }
        val teacher = teacherRepository.findById(teacherId).orElseThrow { RuntimeException("Course or Teacher not found.") }
        course.teacherCourseRelations.add(teacher)
        teacher.teacherCourseRelations.add(course)

        courseRepository.save(course)
        teacherRepository.save(teacher)

        return course
    }

//    @Transactional
//    fun addStudentToCourse(courseId: String, studentId: String): Course {
//        val course = courseRepository.findById(courseId).orElseThrow { RuntimeException("Course not found.") }
//        val student = studentRepository.findById(studentId).orElseThrow { RuntimeException("Student not found.") }
//
////        course.studentCourseRelations.add(student)
//        student.studentCourseRelations.add(course)
//
//        courseRepository.save(course)
//        studentRepository.save(student)
//
//        return course
//    }
}
